app/scripts/lib/app-start.js

```diff
@@ -304,66 +296,35 @@ define(function (require, exports, module) {
 
     initializeAuthenticationBroker () {
       if (! this._authenticationBroker) {
-        let assertionLibrary = this._assertionLibrary;
-        let iframeChannel = this._iframeChannel;
-        let metrics = this._metrics;
-        let oAuthClient = this._oAuthClient;
-        let relier = this._relier;
-        let session = Session;
-        let win = this._window;
-
-        if (this._isFxFirstrunV2()) {
-          this._authenticationBroker = new FxFirstrunV2AuthenticationBroker({
-            iframeChannel,
-            relier,
-            window: win
-          });
-        } else if (this._isFxFirstrunV1()) {
-          this._authenticationBroker = new FxFirstrunV1AuthenticationBroker({
-            iframeChannel,
-            relier,
-            window: win
-          });
-        } else if (this._isFxFennecV1()) {
-          this._authenticationBroker = new FxFennecV1AuthenticationBroker({
-            relier,
-            window: win
-          });
-        } else if (this._isFxDesktopV3()) {
-          this._authenticationBroker = new FxDesktopV3AuthenticationBroker({
-            relier,
-            window: win
-          });
-        } else if (this._isFxDesktopV2()) {
-          this._authenticationBroker = new FxDesktopV2AuthenticationBroker({
-            relier,
-            window: win
-          });
-        } else if (this._isFxDesktopV1()) {
-          this._authenticationBroker = new FxDesktopV1AuthenticationBroker({
-            relier,
-            window: win
-          });
-        } else if (this._isFxiOSV1()) {
-          this._authenticationBroker = new FxiOSV1AuthenticationBroker({
-            relier,
-            window: win
-          });
+        const args = {
+          assertionLibrary: this._assertionLibrary,
+          iframeChannel: this._iframeChannel,
+          metrics: this._metrics,
+          oAuthClient: this._oAuthClient,
+          relier: this._relier,
+          session: Session,
+          window: this._window
+        };
+
+        let context;
+        if (this._isFxDesktopV2()) {
+          if (this._isIframeContext()) {
+            context = Constants.FX_FIRSTRUN_V1_CONTEXT;
+          } else {
+            context = Constants.FX_DESKTOP_V2_CONTEXT;
+          }
         } else if (this._isOAuth()) {
-          this._authenticationBroker = new RedirectAuthenticationBroker({
-            assertionLibrary,
-            metrics,
-            oAuthClient,
-            relier,
-            session,
-            window: win
-          });
+          context = Constants.OAUTH_CONTEXT;
         } else {
-          this._authenticationBroker = new BaseAuthenticationBroker({
-            relier
-          });
+          context = this._getContext();
         }
 
+        const { Constructor, options } = authBrokers.get(context);
+        this._authenticationBroker = new Constructor(options.reduce((options, key) => {
```

Comment:
If it's on the table, I prefer your second option the best: pass everything in and let the constructor just ignore whatever it doesn't need.

I've pushed again with that change added.